src/cache_cf.cc

```diff
@@ -619,17 +627,27 @@ parseConfigFileOrThrow(const char *file_name)
     return err_count;
 }
 
-// TODO: Refactor main.cc to centrally handle (and report) all exceptions.
-int
+bool
 parseConfigFile(const char *file_name)
 {
     try {
-        return parseConfigFileOrThrow(file_name);
+        if (const auto err_count = parseConfigFileOrThrow(file_name)) {
+            // TODO: rollback to previous working settings
+            // for now any errors are a fatal condition...
+            debugs(1, DBG_CRITICAL, "FATAL: Shutdown due to " << err_count << " configuration errors." <<
+                   Debug::Extra << "Log lines above or 'squid -k parse' may have more details.");
+            if (!opt_parse_cfg_only)
+                self_destruct();
+            return false;
```

Comment:
IMO this is still correct to say this debugs message as-is. It is the log output and action that would happen without -k parse.

As documented in the PR description -k parse continues after all parser FATAL, including here.  The only thing we assume is that the caller detect and handle -k parse mode appropriately - no assumption on the _how_ of that.

In fact the reconfigure caller is not reachable by that mode, so does nothing.

Reply:
> IMO this is still correct to say this debugs message as-is. It is the log output and action that would happen without -k parse.

Well, `squid -k parse` is not supposed to output all the same messages as `squid`, especially messages that would be misleading outside of that `squid` context, but we should not bog down in discussing this minor problem. This change request is about the function deciding what its caller will do. A function should not do that without a very good reason, and I see no valid excuses here.

>  we assume is that the caller detect and handle -k parse mode appropriately

Yes, and there is no reason to burden this function with that assumption. It does not help the caller and makes this function logic a lot more complex that it should be.

> In fact the reconfigure caller is not reachable by that mode, so does nothing.

self_destruct() is reachable by the reconfigure caller. This function assumes that the reconfigure caller wants to destroy Squid on errors. That assumption will eventually become wrong, but it is currently wrong for the code to make any such assumption in the first place (as discussed above). Again, I see no need to add any of these problems, current and future.

I am pretty sure that this function should throw instead of calling debugs("FATAL: Shutdown...."), calling self_destruct(), and returning false.

Reply:
> I am pretty sure that this function should throw instead of calling debugs("FATAL: Shutdown...."), calling self_destruct(), and returning false.

Throwing from here will discard all the local variables which need to be used to process the next line by -k parse.

Reply:
> Alex: I am pretty sure that this function should throw instead of calling debugs("FATAL: Shutdown...."), calling self_destruct(), and returning false.

> Amos: Throwing from here will discard all the local variables

"Here", the current PR code already does "discard all the local variables" by calling self_destruct() or returning false. Thus, throwing will not affect that decision to discard.

I am not suggesting something drastic here, just a proper way to implement parsing termination that the existing PR code is already trying to implement. You will need to fix other problems for the whole thing to work well together, of course, but this particular change request is a good first step forward AFAICT.

Reply:
Closing as removal of the return value makes this thread irrelevant to latest code.